### DISCUSSION TITLE
In the Key Exchange API, pointers that are not supposed to be NULL are described as Nullable

### DISCUSSION THREAD

**Discussion by Kleidukos** [jedisct1/libsodium#1171] [Q&A] [answered]
Hi! I'm writing bindings to the Key Exchange API, and I'm reading the signature of `crypto_kx_client_session_keys`:

```c
int crypto_kx_client_session_keys(unsigned char rx[crypto_kx_SESSIONKEYBYTES],
                                  unsigned char tx[crypto_kx_SESSIONKEYBYTES],
                                  const unsigned char client_pk[crypto_kx_PUBLICKEYBYTES],
                                  const unsigned char client_sk[crypto_kx_SECRETKEYBYTES],
                                  const unsigned char server_pk[crypto_kx_PUBLICKEYBYTES]);
```

Now from what I know of the Standard, the array syntax means that these parameters are non-nullable pointers.

Then a little bit down the documentation paragraph, I see this sentence:

> rx and tx are both crypto_kx_SESSIONKEYBYTES bytes long. If only one session key is required, either rx or tx can be set to NULL.

My questions is:

Aren't `tx` and `rx` supposed to be non-nullable?

**Comment by Kleidukos** [comment, marked as answer]
So, I was sorely mistaken, it's actually when the keyword `static` appears in the `[` `]` that the pointer must be non-NULL. 

cf: https://godbolt.org/z/zffd1f7Yn

**Comment by jedisct1** [comment]
Yes, as documented, these parameters can be `NULL`.

Functions with parameters that have to be non-`NULL` are marked with the `__attribute__((nonnull))` attribute in their prototype.

**Reply by Kleidukos** [reply]
Thanks! For now I'm mostly taking the signatures in the docs to document my own bindings to Haskell, but I'll do a sweep in the source code to see if there are any `nonnull` attributes that I would have missed.
